### ISSUE TITLE
Persistent Python wrapper objects

### ISSUE THREAD

**Issue by WMOkiishi** [panda3d/panda3d#1410]
## General Idea
It would be convenient if Panda's C++ objects were consistently exposed to Python with the same instance of a Python wrapper. This would require the C++ object to hold onto a reference to said Python wrapper, though, creating a reference cycle that won't automatically be taken care of.

## First Steps
For cases where the Python wrapper loses all other references (i.e., reaches a reference count of 1) before the C++ object does, the reference cycle can be detected and broken by an overridden `unref` method. It could check if the C++ object and its associated Python wrapper both have reference counts of 1, meaning they must only be referred to by each other. If so, the reference cycle can be broken on the C++ end, allowing both reference counts to reach 0, causing the objects to be cleaned up.

## The Problem
The procedure described above cannot do anything for cases where the C++ object reaches a reference count of 1 before its Python wrapper does. To handle those cases, we will have to hook into one of the following (or possibly some other overlooked area):

- Panda3D's `ReferenceCount::ref`
- Python's [`tp_traverse`](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_traverse)
- Python's [`gc.callbacks`](https://docs.python.org/3/library/gc.html#gc.callbacks)
- Python destructors

## Analysis
Overriding `ref`: If both `ref` and `unref` were overridden, we could decrement the Python wrapper's reference count whenever the C++ reference count reaches 1 and increment it whenever the C++ reference count goes back above 1. This would allow Python to break the reference cycle by destroying the wrapper once it and the C++ object are only referenced by each other. This would require making `ref` virtual, though, incurring a performance penalty. While this could theoretically be worth the cost, other solutions seem more promising for less overhead.

`tp_traverse`: Classes holding onto references to their Python wrappers could implement a custom `tp_traverse` that only reports the wrapper's self-reference while the underlying C++ object's reference count is 1. This essentially allows (and relies on) Python's GC to clean up the wrapper (and thus the reference cycle), but only once we say it's okay to do so. Note that we could still implement the `unref` override as well to clean up the cases it can.

`gc.callbacks`: Python allows for custom callbacks to be called before or after its own garbage collector runs. We could register some function there that checks for any isolated reference cycles of Panda objects and their Python wrappers and cleans them up. Again, this could be done in addition to the `unref` override.

Destructors: It might be possible to not count the C++ object's reference to its Python wrapper, allowing the wrapper's reference count to reach 0, but implement its destructor in a way that doesn't actually get rid of it. When the wrapper is next needed, we would then essentially increment its reference count from 0 back to 1. From Python's perspective, we would just be returning a new object that happens to be stored in the same location as the old one. The memory would be freed once the Python wrapper's reference count is 0 and the C++ object's is 1, likely using an overridden `unref`. A Python [weakref](https://docs.python.org/3/library/weakref.html) callback might also be helpful here.

## Where Next?
This feature has been discussed at length over the years, and I've attempted to summarize the most promising ideas for implementing it here. Hopefully, this works as a good jumping-off-point for the next person wanting to work on it. I imagine the next steps would be discussing anything I've missed and prototyping implementations to get a better idea of their feasibility.

**Comment by rdb** [comment]
Thank you for the excellent summary of the problem and the possible solutions.  To me, the most important reason to do something like this is so we can support Python's GC properly, and detect cycles involving eg. PythonTask objects.  Having support for subclassing Panda objects from Python is a nice-to-have.  To me, persistent wrappers for C++ Panda objects that *aren't* derived in Python--let alone those created by C++ code--is a non-priority.  Do you agree?  I'll explain in a moment why I think this distinction is relevant...

### Storing the persistent wrapper

The thing we haven't really talked about yet is where we store the PyObject* pointer on the C++ end.  We could add this field to some fundamental Panda class like ReferenceCount or TypedObject, but this adds memory usage to *all* C++ objects, even ones that never touch Python.  If we decided to only implement this solution for Panda objects that are created from Python, we could only allocate extra memory for it for these objects, though we'd need to have some way to distinguish whether a given C++ object has this extra memory or not.

Personally, I think we should avoid any solution that introduces a significant overhead to *all* Panda objects; I think the added value of this feature is not great enough to justify that.  That rules out overriding `ref()` in my book.  However, I would be OK with a solution (even GC-based) that only has an overhead for objects that are overridden from Python, including special objects where it is important to support GC such as PythonTask.

I think we can accomplish this by having interrogate automatically construct a subclass type for every possible overrideable class.  Let's say it generates a PyDeriv_PandaNode class that inherits from PandaNode, overrides `unref()`, has a `PyObject *self` member, and we could register it with the C++ typing system.  When creating a new PandaNode, interrogate *actually* constructs a `PyDeriv_PandaNode` if it detects subclassing.  As of fdcfaf69b0b2719a79aee5ca12c093087580db81, we can define a unique "wrap" function for this type which (instead of creating a new Python wrapper object) simply returns the stored "self" field.  Alternatively, we can avoid relying on the C++ typing system by defining a new virtual `get_persistent_python_wrapper()` method on TypedObject that is overridden by our subclass to returns the persistent PyObject*, and have interrogate try that before constructing a new wrapper.

### Breaking the cycles

Then we have to decide how to break the cycles, if not using `ref()`.  I think the destructor trickery-based method might warrant some further looking into, since it seems to be the solution with the least amount of overhead, compared to relying on GC to break the cycle.

I did a bit of research today and learned that since [PEP 442](https://peps.python.org/pep-0442/) it is permissible for a finalizer (`__del__`, or weakref callback) to "resurrect" an object by incrementing its refcount.  So maybe we can register a finalizer with something that does this:
* Decrements the C++ refcount, checking whether it reached 0.
* If so, we let the Python object destruct as well.
* If not, increments the Python reference count to 1, which cancels the destruction (and incidentally also makes it unreachable to the GC).
* It also sets a flag indicating that the object is in "stasis mode".

Now that the Python wrapper object is in stasis mode, the C++ object is allowed to destruct, which would trigger the actual destruction of the Python object as well.  But if in the meantime a new Python reference is created, then we check whether the object is in "stasis" mode, and if so, take the object out of "stasis" mode, which means re-incrementing the C++ reference count and decrementing (or rather: not incrementing) the Python reference count, and clearing the "stasis" flag.

The only problems with this I can see are (1) other weakref callbacks may get run when the object enters stasis mode, which can be confusing if users have their own weakref callbacks, and (2) our finalizer is called only *once*, though there might be a way around this, such as by re-registering the finalizer when take the object out of stasis, or something of the sort.

If we went this route, I would personally prefer supporting it only for subclasses so we can simply override `tp_finalize` and not have to register a weakref callback.

**Comment by WMOkiishi** [comment]
> To me, persistent wrappers for C++ Panda objects that aren't derived in Python--let alone those created by C++ code--is a non-priority. Do you agree?

I am inclined to agree, yes. I don't think preserving the wrapper's object ID is important, so we would only need to be concerned with wrappers that have custom Python subclasses or attributes. If we were to use the a custom subclass as a sort of "flag" to preserve the instance, that should cover most uses--setting custom attributes on a Panda wrapper usually requires a subclass anyways. However, it seems that `PythonTask` does already support custom Python attributes, so it might be a good idea to extend the system to that, too, if it's simpler or more capable than whatever is currently being used. (It looks like you might've been thinking the same thing.)

Your ideas for storing the persistent wrapper seem good to me. Other than nodes and tasks, are there any other Panda wrappers a user might want to subclass? I personally can't think of any, but I also haven't used Panda very extensively.

Object resurrection borders on the limits of my programming knowledge, and is certainly outside what I have experience with. But I do agree that it could be the most straightforward solution. Perhaps it would be helpful to know the intended use for object resurrection in Python in order to get a better idea of how far outside it (if at all) we'd be straying.

**Comment by rdb** [comment]
> Other than nodes and tasks, are there any other Panda wrappers a user might want to subclass?

No more than I can think of reasons to subclass PandaNode.  If we support it for PandaNode we could support it for any TypedReferenceCount subclass, though, especially if our approach has minimal overhead.

I know NodePath is extensively subclassed but since NodePath isn't a reference-counted class and is a lightweight handle to a node (and not a node itself) it makes a lot less sense to uniquify all NodePaths pointing to the same node - we *could* theoretically support this in some elaborate fashion, but I don't personally think this is worth it.

**Comment by WMOkiishi** [comment]
I've had another idea on this. Instead of storing the entire Python wrapper object to return later, we might be able to instead store an intermediate object holding just the data we want to keep. The C++ object would interact with this intermediate object the way we've discussed it interacting with the wrapper, using `unref` to clean up if they both only reference each other. Python wrappers would reference (but not be referenced by) the intermediate object, accessing and storing data on it instead of themselves. As such, we would be free to continue to let them be garbage-collected by Python. The `tp_finalize` function for the wrapper would then check the same conditions `unref` does on the C++ side and clean things up if necessary. This means the condition would be checked regardless of which language stops using an object first.

I think this approach would work, but my limited knowledge of Python extensions means I don't know exactly how it would be implemented in practice. In particular, the structure of the intermediate object is mostly a mystery to me. I'd like to further examine the practicality of the approach, though, so thoughts and feedback are welcome.

**Comment by rdb** [comment]
I took a step towards this in 115a716a7dd30a82ed3c3691210388c3f11d4a8c, which (among other things) adds a way to register a custom wrapping function for a TypeHandle, which interrogate invokes instead of constructing a new Python wrapper.  This was done so that TypedWritable subclasses could have persistent PyObject pointers.

In that use case the reference cycle wasn't a problem because TypedWritable isn't reference-counted on the C++ end.  We still have to solve this to make it work for types like PythonTask and PandaNode and subclasses thereof.  In the current design, subclasses would have to be registered with the typing system, but this could perhaps be done automatically.

I think messing with Python implementation tricks, weak references or object resurrection is a minefield, especially with their recent trend of closing down the internals, and also with the object structure undergoing changes in Python 3.11 and 3.12.  I'm leaning towards either relying on GC (implementing tp_traverse), or doing something like this:
- When C++ objects reach refcount 1, check whether the Python refcount is > 1
  - If so, add the C++ object to a stasis list
  - If not, drops the reference to the Python object, letting both destruct
- A GC callback goes through the stasis list and checks whether the Python refcount is still > 1
  - If so, do nothing
  - If not, remove from the stasis list and break the cycle

That said, I don't see much of a practical difference of this approach over the tp_traverse one, since in both cases the reference is only cleared on a traversal, and we can still in both cases override `unref()` to clean things up when the C++ refcount reaches 1 without waiting for a traversal.

**Comment by rdb** [comment]
I implemented this in #1640, going down the `tp_traverse` route.  I wanted to go down the callback route initially, but I realised this won't work (for reasons I unfortunately no longer remember).  We can still add a custom garbage collection function, or force a more specific traversal to occur at another time, if we so desire.

There might still be an unsolved edge case regarding weak references, which may cause the refcount to go back up in another thread while we're breaking up the cycle.  I haven't fully reasoned through whether this can be an issue, nor do I know how to solve this right now.  I think that in the worst case, it will not cause significant harm, but simply cause the Python wrapper object to be lost prematurely while the C++ object is still alive.
